core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+//! Generate a multiproof from a vector of path proofs.
+//! The multiproof will contain the minimum information needed to verify
+//! the inclusion of all provided path proofs.
+
+use crate::{
+    proof::{PathProof, PathProofTerminal},
+    trie::Node,
+};
+use bitvec::{order::Msb0, view::BitView};
+
+// Each terminal node may have a set of siblings that are uniquely
+// used during the verification with its terminal node.
+// This struct contains the terminal node itself,
+// the depth at which the siblings start to be uniquely associated to that terminal,
+// and the siblings themselves.
+#[derive(Debug, Clone)]
+struct SubPathProof {
+    terminal: PathProofTerminal,
+    depth: u8,
+    inner_siblings: Vec<Node>,
+}
+
+/// A proof of multiple path through the trie.
+#[derive(Debug, Clone)]
+pub struct MultiProof {
+    // All subpaths related to a single terminal node
+    sub_paths: Vec<SubPathProof>,
+    // Vector containing the minimum number of nodes required
+    // to reconstruct all other nodes later.
+    external_siblings: Vec<Node>,
+}
+
+impl MultiProof {
+    /// Construct a MultiProof from a vector of *ordered* PathProof
+    pub fn from_path_proofs(path_proofs: Vec<PathProof>) -> Self {
+        // A multi-proof can be viewed by associating each terminal node
+        // with its first n uniquely related siblings from its path proof,
+        // followed by all necessary siblings that are not derivable from
+        // the previously mentioned siblings.
+        let mut sub_paths: Vec<SubPathProof> = vec![];
+        let mut external_siblings: Vec<Node> = vec![];
+
+        // The goal is to traverse the entire tree
+        // formed by all path proofs and only collect the siblings
+        // that cannot be reconstructed in the future
+        //
+        // The traversal does not occur on the siblings themselves,
+        // but on the ordered set of key_paths within PathProofs.
+        //
+        // For example, take two key_paths starting from the first bit.
+        // If two keys share the same bit at index i, they will have the same sibling at index i.
```

Comment:
If two keys have the same prefix of 'i' bits, they will have the same sibling at index i.

As written, this says `010` and `110` both have the same sibling at index 2.

Reply:
oh yes, that's wrong, it should probably be:

If two keys share bits up to index i, they will have the same sibling at index i.